### PR TITLE
Setup middleware to handle Modal exceptions and improve logging

### PR DESCRIPTION
**Pull request by hholb**
Related: #187 -- does not resolve the ticket completely

## Overview

This PR sets us up for better error reporting to the frontend when there are issues with modal deployment/invocations that users should know about, as well as improving the error logs coming from the fastapi server.

This PR defines a new custom exception type `ModalException` that includes fields for:
- details about the error
- suggested fix for the user if it is something they can fix in their code
- http status code

There is logic added to the exception handling middleware that converts a `ModalException` to a `JSONResponse` before sending the response to the client. The middleware logs important details about the exception as well as a stack trace so we can have a better idea of how to help users when something goes wrong. A big improvement in the logging comes from a helper function to remove excess noise from the stack trace.

Since the modal validation/deployment/invocation routes are still in flux, I figured getting a scaffolding in place so that we can raise a `ModalException` with a suggested fix for the user is a good first step. This PR doesn't include errors messages for everything that could wrong, just gets a system in place so we can write specific error messages as we get stuff ironed out. I added a couple `ModalException`s into the `add_modal_app` route so we can see how this setup works in context, however I expect these errors will change as the `add_modal_app` route gets tweaked.

## Discussion

I am not sure that defining the `ModalException` in the dependencies module makes the most sense, but I haven't thought of a better place. I am open to suggestions.

## Testing

Manual testing

### CODE REVIEW

File: garden-backend-service/src/api/dependencies/modal.py [hunk 1]
```diff
@@ -8,3 +9,28 @@ async def get_modal_client(settings: Settings = Depends(get_settings)) -> modal.
     return await modal.client._Client.from_credentials(
         settings.MODAL_TOKEN_ID, settings.MODAL_TOKEN_SECRET
     )
+
+
+class ModalException(Exception):
```
Comment (OwenPriceSkelly) [inline garden-backend-service/src/api/dependencies/modal.py]: We won't be using these with `Depends` anywhere so I don't think this is the right place for these to live -- maybe in a new `src.exceptions` module?

File: garden-backend-service/src/middleware/logging.py [hunk 2]
```diff
@@ -1,11 +1,43 @@
...
 import structlog
-from fastapi import HTTPException, Request, Response
+from fastapi import Request, Response
 from fastapi.responses import JSONResponse
 from starlette.middleware.base import BaseHTTPMiddleware
 
+from src.api.dependencies.modal import ModalException, handle_modal_exception
+
+
+def filter_stacktrace(exc: Exception, func_to_stop_at: str = "run_endpoint_function"):
```
Comment (OwenPriceSkelly) [inline garden-backend-service/src/middleware/logging.py]: nit: can we name this helper function something like `format_traceback` or `process_traceback` to indicate that it's building a nicely cleaned up string for us and not a modified traceback object or list of `FrameSummary` objects or anything

Comment (OwenPriceSkelly) [reply garden-backend-service/src/middleware/logging.py]: I also think we might want this helper to live in `src.exceptions`

**Review by OwenPriceSkelly — APPROVED** [review]
LGTM! couple superficial nits about where the new code should live but otherwise this looks great 🙌

File: garden-backend-service/src/api/dependencies/modal.py (hunk 1 above)
Comment (hholb) [reply garden-backend-service/src/api/dependencies/modal.py]: Good idea! New module incoming!

**Pull request merged**